### PR TITLE
Issue #469:  support training restorage

### LINKED ISSUE #469: Support training restorage
## Subject of the feature

Currently, we are using `model.load_weights` to start the supervised training from non-random weights.
However, if the intention is to resume a stopped run, this will not be perfect as the status of optimizer is not saved in the checkpoint.

The following piece of code can be helpful:
https://github.com/tensorflow/tensorflow/issues/27861#issuecomment-487455939

### PR DESCRIPTION
**Pull request by kate-sann5100**
# Description

Please include a summary of the change and which issue is fixed. Please also include
relevant motivation and context. List any dependencies that are required for this
change.

Fixes #469 

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [x] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: deepreg/callback.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        model, log_dir + "/save", period=save_period
+    )
+    return [tensorboard_callback, checkpoint_manager_callback]
+
+
+class CheckpointManagerCallback(tf.keras.callbacks.Callback):
+    """
+    Callback wrapping `tf.train.CheckpointManager`.
+
+    :param model: model
```
Comment (mathpluscode) [inline deepreg/callback.py:26]: docstring regarding params should be under init

File: deepreg/callback.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    :param directory: directory to store the checkpoints
+    :param period: save the checkpoint every X epochs
+    :param save_on_train_end: save the checkpoint as the training ends
+    """
+
+    def __init__(self, model, directory, period=1, save_on_train_end=True):
```
Comment (mathpluscode) [inline deepreg/callback.py:32]: `perido:int=1, save_on_train_end:bool=True`

File: test/unit/test_callback.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                new_reader.get_tensor(k)
+            )
+            assert equal.all(), k
+
+    # remove temporary ckpt directories
+    shutil.rmtree("./test/unit/ckpt_old")
+    shutil.rmtree("./test/unit/ckpt_new")
+
+
+# if __name__ == '__main__':
```
Comment (mathpluscode) [inline test/unit/test_callback.py:67]: remove this

File: test/unit/test_callback.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import shutil
+
+import numpy as np
+import tensorflow as tf
+
+from deepreg.callback import CheckpointManagerCallback
+
+
+def test_restore_CheckpointManagerCallback():
```
Comment (mathpluscode) [inline test/unit/test_callback.py:9]: rename to test_restore_checkpoint_manage_callback

**Comment by kate-sann5100** [comment]
@mathpluscode @NMontanaBrown @YipengHu Anyone have time to review the merge request please?

**Review by YipengHu — APPROVED** [review]

**Comment by NMontanaBrown** [comment]
I can review today!

> @mathpluscode @NMontanaBrown @YipengHu Anyone have time to review the merge request please?

**Review by mathpluscode — APPROVED** [review]

**Pull request merged**